File: csharp/Huffman.cs
using System;
using System.Collections.Generic;
using System.Linq;

public class HuffmanNode : IComparable<HuffmanNode>
{
    public char? Char { get; set; }
    public int Frequency { get; set; }
    public HuffmanNode Left { get; set; }
    public HuffmanNode Right { get; set; }

    public HuffmanNode(char? ch, int freq)
    {
        Char = ch;
        Frequency = freq;
    }

    public int CompareTo(HuffmanNode other)
    {
        return Frequency.CompareTo(other.Frequency);
    }
}

public class HuffmanCoding
{
    private HuffmanNode Root;
    public Dictionary<char, string> Codes = new();
    public Dictionary<string, char> ReverseCodes = new();

    private Dictionary<char, int> BuildFrequencyMap(string text)
    {
        var freqMap = new Dictionary<char, int>();
        foreach (char c in text)
        {
            if (!freqMap.ContainsKey(c))
                freqMap[c] = 0;
            freqMap[c]++;
        }
        return freqMap;
    }

    private HuffmanNode BuildHuffmanTree(Dictionary<char, int> freqMap)
    {
        var heap = new List<HuffmanNode>();
        foreach (var kv in freqMap)
            heap.Add(new HuffmanNode(kv.Key, kv.Value));

        heap.Sort();

        while (heap.Count > 1)
        {
            heap.Sort();
            var left = heap[0];
            var right = heap[1];
            heap.RemoveRange(0, 2);

            var merged = new HuffmanNode(null, left.Frequency + right.Frequency)
            {
                Left = left,
                Right = right
            };

            heap.Add(merged);
        }

        Root = heap[0];
        return Root;
    }

    private void GenerateCodes(HuffmanNode node, string currentCode)
    {
        if (node == null) return;

        if (node.Char != null)
        {
            Codes[node.Char.Value] = currentCode;
            ReverseCodes[currentCode] = node.Char.Value;
            return;
        }

        GenerateCodes(node.Left, currentCode + "0");
        GenerateCodes(node.Right, currentCode + "1");
    }

    public (string encodedText, Dictionary<char, string> codes, Dictionary<char, int> frequencyMap) Encode(string text)
    {
        var freqMap = BuildFrequencyMap(text);
        BuildHuffmanTree(freqMap);
        GenerateCodes(Root, "");

        var encoded = string.Concat(text.Select(c => Codes[c]));

        return (encoded, Codes, freqMap);
    }

    public string Decode(string encodedText, Dictionary<char, string> codes = null)
    {
        if (codes != null)
        {
            ReverseCodes.Clear();
            foreach (var kv in codes)
                ReverseCodes[kv.Value] = kv.Key;
        }

        string currentCode = "";
        var decoded = new List<char>();

        foreach (char bit in encodedText)
        {
            currentCode += bit;
            if (ReverseCodes.ContainsKey(currentCode))
            {
                decoded.Add(ReverseCodes[currentCode]);
                currentCode = "";
            }
        }

        return new string(decoded.ToArray());
    }
}

File: python/huffman.py
import heapq

class HuffmanNode:
    def __init__(self, char, frequency):
        self.char = char
        self.frequency = frequency
        self.left = None
        self.right = None

    def __lt__(self, other):
        return self.frequency < other.frequency


class HuffmanCoding:
    def __init__(self):
        self.root = None
        self.codes = {}
        self.reverse_codes = {}

    def build_frequency_map(self, text: str) -> dict:
        frequency_map = {}
        for char in text:
            frequency_map[char] = frequency_map.get(char, 0) + 1
        return frequency_map

    def build_huffman_tree(self, frequency_map: dict):
        heap = [HuffmanNode(char, freq) for char, freq in frequency_map.items()]
        heapq.heapify(heap)

        while len(heap) > 1:
            left = heapq.heappop(heap)
            right = heapq.heappop(heap)
            merged = HuffmanNode(None, left.frequency + right.frequency)
            merged.left, merged.right = left, right
            heapq.heappush(heap, merged)

        self.root = heap[0] if heap else None
        return self.root

    def generate_codes(self, node=None, current_code=""):
        if node is None:
            node = self.root
        if node is None:
            return

        if node.char is not None:
            self.codes[node.char] = current_code
            self.reverse_codes[current_code] = node.char
            return

        self.generate_codes(node.left, current_code + "0")
        self.generate_codes(node.right, current_code + "1")

    def encode(self, text: str):
        freq_map = self.build_frequency_map(text)
        self.build_huffman_tree(freq_map)
        self.generate_codes()

        encoded = "".join(self.codes[char] for char in text)
        return {
            "encodedText": encoded,
            "codes": self.codes,
            "frequencyMap": freq_map,
        }

    def decode(self, encoded_text: str, codes=None) -> str:
        if codes and isinstance(codes, dict):
            self.reverse_codes = {code: char for char, code in codes.items()}

        current_code = ""
        decoded = []
        for bit in encoded_text:
            current_code += bit
            if current_code in self.reverse_codes:
                decoded.append(self.reverse_codes[current_code])
                current_code = ""
        return "".join(decoded)
